### ISSUE TITLE
Will not compile under LuaJIT with strict globals

### ISSUE THREAD

**Issue by Olical** [bakpakin/Fennel#331]
Running `make` under LuaJIT (Lua 5.1) fails due to the following error:

```
Compile error in 'rawlen' src/fennel/specials.fnl:837: unknown global in strict mode: rawlen
```

I'm not sure if compiling the project under LuaJIT is a goal or not but I thought it was interesting. Removing the reference in `specials.fnl` fixes the issue but I'll be swapping to Lua 5.2+ (where `rawlen` exists) going forward. I just wanted to bring it to your attention just in case it wasn't intended and should be fixed.

**Comment by Olical** [comment]
I should open this as another issue potentially but I've also been noticing this error when loading Fennel into Neovim (LuaJIT) _very_ intermittently:

```
E5105: Error while calling lua chunk: ...e/ollie/repos/Olical/aniseed/lua/aniseed/deps/fennel.lua:2335: attempt to index a nil value
```
Worth opening another issue for that? The code in question:

```lua
    local function parse_stream()
      local whitespace_since_dispatch, done_3f, retval = true
      local function dispatch(v)
        if (#stack == 0) then
          retval, done_3f, whitespace_since_dispatch = v, true, false
          return nil
        elseif stack[#stack].prefix then # -- Here.
          local stacktop = stack[#stack]
          stack[#stack] = nil
          return dispatch(utils.list(utils.sym(stacktop.prefix), v))
        else
          whitespace_since_dispatch = false
          return table.insert(stack[#stack], v)
        end
      end
      local function badend()
        local accum = utils.map(stack, "closer")
        return parse_error(string.format("expected closing delimiter%s %s", (((#stack == 1) and "") or "s"), string.char(unpack(accum))))
      end
...
```

**Comment by technomancy** [comment]
The first problem is just because the Fennel compiler is intended to be compiled with strict globals checking off, regardless of what Lua version it's compiled under. It wouldn't hurt to make it work with strict checking on, but it's just never come up because normally when you use Fennel to compile Fennel itself it's done as AOT, which disables strict checks.

The other problem is definitely separate and worth another report with a repro case.

**Comment by Olical** [comment]
Ah okay! I'll try again with this turned off if I can work out how. And I'll see if I can repro this race condition in a minimal way. My gut says I won't be able to and it's only in the context of Neovim and my tooling that I see it, but we'll see! Will report back in this week(end).

**Comment by Olical** [comment]
By the way, I'm not turning strict global checking on in any way and I'm not sure how it could be turned off :thinking: I can't find anything through a cursory Google but I'll take a deeper look soon. The rare startup error is more worrying for me right now anyway so that's the main one I'd like to repro. I'll just rely on non JIT Lua for compiling Fennel for now.

**Comment by technomancy** [comment]
> I'm not turning strict global checking on in any way and I'm not sure how it could be turned off

Ah right--to be clear, strict global checking is on by default everywhere other than during AOT compilation. You can turn it off elsewhere by passing `:allowedGlobals false` in the opts table used by the compiler, (I realized this was missing from the API docs, so I'll add that) or you could just remove the stray global from your bundled copy of the compiler.

**Comment by technomancy** [comment]
Got a fix for this in e685d78.

**Comment by Olical** [comment]
I can reproduce the race condition by loading Fennel in LuaJIT from my shell (fish, but still).

```
for i in (seq 100)                                                                                  develop aniseed
        lua -e "require('lua.aniseed.deps.fennel')"
    end
lua: ./lua/aniseed/deps/fennel.lua:2335: attempt to index a nil value
stack traceback:
	./lua/aniseed/deps/fennel.lua:2335: in function 'dispatch'
	./lua/aniseed/deps/fennel.lua:2409: in function '(for generator)'
	./lua/aniseed/deps/fennel.lua:1965: in function 'compile-string'
	./lua/aniseed/deps/fennel.lua:2797: in function 'eval'
	./lua/aniseed/deps/fennel.lua:3152: in main chunk
	[C]: in function 'require'
	(command line):1: in main chunk
	[C]: at 0x004061d0
lua: ./lua/aniseed/deps/fennel.lua:2335: attempt to index a nil value
stack traceback:
	./lua/aniseed/deps/fennel.lua:2335: in function 'dispatch'
	./lua/aniseed/deps/fennel.lua:2409: in function '(for generator)'
	./lua/aniseed/deps/fennel.lua:1965: in function 'compile-string'
	./lua/aniseed/deps/fennel.lua:2797: in function 'eval'
	./lua/aniseed/deps/fennel.lua:3152: in main chunk
	[C]: in function 'require'
	(command line):1: in main chunk
	[C]: at 0x004061d0
```

If I try to observe the issue with if statements and printing it goes away. I have a horrible feeling `#stack` is returning `nil` sometimes (which triggers my debug if statement) but on printing, the second call to `#stack` is `0`. So I _think_ it's a heisenbug where observing a `#stack` of `nil` causes the next lookup to be `0` which causes the `stack[#stack].prefix` to try and lookup `prefix` in a `nil`.

Want me to open a issue with more background? I guess you'd like to see this with your own eyes to prove it's not just me before that?

**Comment by Olical** [comment]
Minimal bash example:

```bash
$ for i in {1..100}; do lua -e "require 'fennel'"; done
lua: ./fennel.lua:2335: attempt to index a nil value
stack traceback:
	./fennel.lua:2335: in function 'dispatch'
	./fennel.lua:2409: in function '(for generator)'
	./fennel.lua:1965: in function 'compile-string'
	./fennel.lua:2797: in function 'eval'
	./fennel.lua:3152: in main chunk
	[C]: in function 'require'
	(command line):1: in main chunk
	[C]: at 0x004061d0
lua: ./fennel.lua:2335: attempt to index a nil value
stack traceback:
	./fennel.lua:2335: in function 'dispatch'
	./fennel.lua:2409: in function '(for generator)'
	./fennel.lua:1965: in function 'compile-string'
	./fennel.lua:2797: in function 'eval'
	./fennel.lua:3152: in main chunk
	[C]: in function 'require'
	(command line):1: in main chunk
	[C]: at 0x004061d0
```

**Comment by Olical** [comment]
If I add

```lua
        if stack[#stack] == nil and #stack ~= 0 then
          print("===", #stack, stack[#stack], stack)
        end
```

At line 2332 (just before the if statement that contains this error) it no longer happens. I'm so sorry for finding this :grimacing:

**Comment by technomancy** [comment]
> $ for i in {1..1000}; do luajit -e "require 'fennel'"; done

I've run a few thousand iterations of this both on main and 0.6.0, but it never fails.

I'm on LuaJIT 2.0.4; maybe it needs a specific LuaJIT version to trigger the bug?

**Comment by technomancy** [comment]
Regardless of whether the problem is caused by a specific LuaJIT issue, the function in question is odd and unidiomatic, probably because it was compiled from the old Lua version of Fennel using Antifennel. I've refactored it into a slightly nicer version on the `parse-stream-tweak` branch; you could see if that runs into the same issue?

https://git.sr.ht/~technomancy/fennel/commit/parse-stream-tweak

**Comment by Olical** [comment]
Sure thing! The LuaJIT I saw the issue on was `LuaJIT 2.1.0-beta3`, definitely the right `fennel.lua` since I could alter the source and see the changes reflected. It was with this version of the Fennel repo `69675424a9c8e28d999d56754d00796eb49dafb9`.

I was seeing this issue in a fresh clone of the Fennel repo, detached from Aniseed entirely, it was about as clean as I could get it. Also I'm on NixOS with a m2 SSD and generally high spec machine. I do wonder if it took some very very specific circumstances to appear.

I can however say that `parse-stream-tweak` is totally fine! I can't reproduce the issue after 1000's of reads when I can consistently reproduce it about 2% of the time on `69675424a9c8e28d999d56754d00796eb49dafb9`.

You're the best :smile:

**Comment by jaawerth** [comment]
@Olical  that did it! Looks like what we've got here, I think, an upstream bug in LuaJIT 2.1-beta3. I was able to reproduce the behavior you're seeing after running the same check in a freshly built LuaJit from their `2.1` branch. Note that tag `2.1-beta3` in their git repo is older and did not reproduce the issue, but he current 2.1 branch still reports as that same version.

I believe it should be impossible for `stack[#stack]` to ever be nil, since once you set it to nil, `#stack` will report as one less. This to me supports the idea it's probably a vm bug, which we'll probably want to submit upstream.
